Source/AssetRipper.Processing/AnimationClips/AnimationClipConverter.cs

```diff
@@ -53,111 +53,123 @@ private void ProcessInner()
-			float interval = 1.0f / sampleRate;
 
-			// first (index [0]) stream frame is for slope calculation for the first real frame (index [1])
-			// last one (index [count - 1]) is +Infinity
-			// it is made for slope processing, but we don't need them
-			bool frameIndex0 = true;
-			for (int frameIndex = 0; frameIndex < streamFrames.Count - 1; frameIndex++)
+			if (streamedFrames.Count > 1)
 			{
-				StreamedFrame frame = streamFrames[frameIndex];
-				for (int curveIndex = 0; curveIndex < frame.Curves.Length;)
-				{
-					StreamedCurveKey curve = frame.Curves[curveIndex];
-					IGenericBinding binding = bindings.FindBinding(curve.Index);
+				streamedFrames[0].Time = 0f; // fix first frame for PPtrCurves to have Time=0 instead of float.MinValue
+			}
+			int frameCount = streamedFrames.Count - 1; // last StreamedFrame is dummy so must be skipped
+			for (int frameIdx = 0; frameIdx < frameCount; frameIdx++)
+			{
+				// last real frame doesn't need outSlope calculation, and will have inSlope from previous iteration
+				bool doSlopeCalc = frameIdx != frameCount - 1;
 
+				StreamedFrame frame = streamedFrames[frameIdx];
+				for (int curveIdx = 0; curveIdx < frame.Curves.Length;)
+				{
+					int curveID = frame.Curves[curveIdx].Index;
+					IGenericBinding binding = GetBinding(curveID);
 					string path = GetCurvePath(binding.Path);
+					StreamedCurveKey curve;
 					if (binding.IsTransform())
 					{
-						if (frameIndex0)
+						int transformDim = binding.TransformType().GetDimension();
+						if (frameIdx == 0) // first StreamedFrame only contains PPtrCurves, skip
 						{
-							curveIndex = GetNextCurve(frame, curveIndex);
+							curveIdx += transformDim;
 							continue;
 						}
-						GetPreviousFrame(streamFrames, curve.Index, frameIndex, out int prevFrameIndex, out int prevCurveIndex);
-						int dimension = binding.TransformType().GetDimension();
-						for (int key = 0; key < dimension; key++)
+						for (int offset = 0; offset < transformDim; offset++)
 						{
-							StreamedCurveKey keyCurve = frame.Curves[curveIndex];//index out of bounds
-							StreamedFrame prevFrame = streamFrames[prevFrameIndex];
-							StreamedCurveKey prevKeyCurve = prevFrame.Curves[prevCurveIndex + key];
-							float deltaTime = frame.Time - prevFrame.Time;
-							curveValues[key] = keyCurve.Value;
-							inSlopeValues[key] = prevKeyCurve.CalculateNextInSlope(deltaTime, keyCurve.Value);
-							outSlopeValues[key] = keyCurve.OutSlope;
-							curveIndex = GetNextCurve(frame, curveIndex);
+							curve = frame.Curves[curveIdx];
+							if (doSlopeCalc)
+							{
+								if (TryGetNextFrame(streamedFrames, frameIdx, curveID, out StreamedFrame? nextFrame, out int nextCurveIdx))
+								{
+									StreamedCurveKey nextCurve = nextFrame.Curves[nextCurveIdx + offset];
+									curve.CalculateSlopes(frame.Time, nextFrame.Time, nextCurve);
+								}
+							}
+							curveValues[offset] = curve.RightSidedLimit;
```

Comment:
On second thought, it's not immediately clear why `RightSidedLimit` is used here. I think we should have a new property:
```cs
public float Value { get => RightSidedLimit; private set => RightSidedLimit = value; }
```
`Value` and `RightSidedLimit` should each be used where they intuitively make sense.

Reply:
implemented